### ISSUE TITLE
Passenger 5.1.8 no longer starts due to log file permissions

### ISSUE THREAD

**Issue by graaff** [phusion/passenger#1990]
Passenger 5.1.11 no longer starts on some of my apache installations due to a permission issue with the error_log log file:

As mentioned in the error_log:

```
[ E 2017-10-16 18:34:46.0165 12101/T1 age/Sha/Fun/Initialization.cpp:569 ]: *** ERROR: Cannot open /var/log/apache2/error_log for writing: Permission denied (errno=13)
     in 'Passenger::VariantMap Passenger::Agent::Fundamentals::initializeAgent(int, char***, const char*, Passenger::Agent::Fundamentals::OptionParserFunc, Passenger::Agent::Fundamentals::PreinitializationFunc, int)' (Initialization.cpp:538)

[Mon Oct 16 18:34:47.023013 2017] [passenger:error] [pid 12095:tid 139929927526272] *** Passenger could not be initialized because of this error: Unable to start the Phusion Passenger watchdog because it encountered the following error during startup: Unable to start the Passenger core: it seems to have crashed during startup for an unknown reason, with exit code 1

```

This causes both the watchdog and core not to start, and thus to Passenger not running at all.

The issue appears to be tied to the PassengerUserSwitching setting. When this is On the error does not occur and Passenger starts as expected. When PassengerUserSwitching is Off the permission issue arises.

In both cases PassengerDefaultUser is set to apache and the error_log is "-rw-r--r-- 1 root root".

**Question 2: Passenger version and integration mode**:

Passenger open source 5.1.11 apache module (with apache 2.4.27)

**Question 3: OS or Linux distro, platform (including version)**:

Gentoo Linux (amd64). I am the package maintainer there.

**Question 4: Passenger installation method**:

Gentoo Linux package

**Question 5: Your app's programming language (including any version managers) and framework (including versions)**:

ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux]

**Question 6: Are you using a PaaS and/or containerization? If so which one?**

No

**Comment by graaff** [comment]
After doing some more testing it turns out that this issue was already present in passenger 5.1.8. Downgrading to 5.1.6 makes it go away on the same setup. We never packaged 5.1.7 on Gentoo so I can't easily test this. Let me know if this is important. In any case this seems to point to the log file changes in 5.1.8:

 * Introduces a shorter and more informative log format. The new format significantly shortens the thread ID, and includes the message's log level (critical, error, warning, notice, etc) so that users can more easily filter out uninteresting messages.

**Comment by graaff** [comment]
I have determined, as I suspected, that Passenger tries to open the log file as the "apache" user in my case.

**Comment by CamJN** [comment]
It seems like when we dup the stderr file descriptor the file needs to be re-opened for some reason. Is there some peculiarity in Gentoo regarding FD inheritance?

**Comment by graaff** [comment]
Not that I'm aware of but I'll ask around. I initially saw this on our hardened servers with the grsecurity patch set applied to the kernel, but I can also reproduce this on the standard Gentoo Linux kernel (which is really the upstream kernel with very little patching) with no additional security measures. This is with 4.9.49-gentoo-r1.

**Comment by graaff** [comment]
Here what looks like the relevant parts of an strace session:


```
30425 open("/var/log/apache2/error_log", O_WRONLY|O_CREAT|O_APPEND, 0644) = 8
30425 dup2(8, 2)                        = 2
```

Later at what looks like the Passenger initialization, the file is simply opened, no sign of dup2 here.
```
30428 execve("/usr/lib64/passenger/support-binaries/PassengerAgent", ["PassengerAgent", "core", "                                "...], [/* 85 vars */]) = 0
... snip ...

30428 stat("/usr/lib64/ruby/vendor_ruby/phusion_passenger/locations.ini", {st_mode=S_IFREG|0644, st_size=584, ...}) = 0
30428 open("/usr/lib64/ruby/vendor_ruby/phusion_passenger/locations.ini", O_RDONLY) = 8
30428 read(8, "[locations]\npackaging_method=deb"..., 8191) = 584
30428 read(8, "", 8191)                 = 0
30428 close(8)                          = 0
30428 open("/var/log/apache2/error_log", O_WRONLY|O_CREAT|O_APPEND, 0644) = -1 EACCES (Permission denied)
30428 futex(0x7f880368f1a0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
30428 open("/etc/localtime", O_RDONLY|O_CLOEXEC) = 8
```

**Comment by CamJN** [comment]
Ah yeah, I see the problem. Ref so it's easy to work on: https://github.com/phusion/passenger/blob/stable-5.1/src/cxx_supportlib/LoggingKit/Implementation.cpp#L661-L672

**Comment by CamJN** [comment]
thanks @graaff

**Comment by graaff** [comment]
No problem. Let me know if you need me to test a patch.

**Comment by FooBarWidget** [comment]
I've fixed this issue in the `loggingkit_initial_open_fix` branch, now pending review. The cause of the issue is as follows.

When user switching is turned off, the Passenger core runs as a non-root user. Both in 5.17 and in 5.1.8, the Passenger core tries to open the log file during startup. But in 5.1.7 it silently ignores the open error and falls back to writing to stderr (which Apache had already redirected to the Apache log file). In 5.1.8, the open failure is no longer silent and causes an abort.

My fix works as follows. We open the log file inside Apache and Nginx and ensure that stderr is redirected to that. Then we make the Passenger Watchdog and the Passenger Core use stderr directly instead of having them open the log file separately.

Now that I think about it, this fix also happens to fix another bug which was present in 5.1.7: if the user had configured `PassengerLogFile` (Apache) / `passenger_log_file` (Nginx), then if the Passenger Core is unable to open it, all logs simply ended up in the Apache/Nginx error log with no message saying why.

**Comment by graaff** [comment]
Fix confirmed on my test system. 👍
